### DISCUSSION TITLE
Proposal: Multiple optional dependencies

### DISCUSSION THREAD

**Discussion by sirmonti** [composer/composer#11828] [General]
The idea is to provide a set of related dependencies and composer installs one of them if none of them are installed.
Example, I wrote a program that needs an PSR7 message class and the program supports this (In order of preference):

- httpsoft/http-message
- nyholm/psr7
- guzzle/psr7
- laminas/laminas-diactoros
- slim/psr7

The idea is, composer verify if one of this packages is installed, if yes, it does nothing, if not, it try to install first package, if it can be installed because conflicts, go to the next, until found one installable. If none is installable, composer will report an unresolvable dependency error.

**Comment by Seldaek** [comment]
The way to do this IMO is to require one of the virtual packages like:

[psr/http-factory-implementation](https://packagist.org/providers/psr/http-factory-implementation): 1.0
[psr/http-message-implementation](https://packagist.org/providers/psr/http-message-implementation): 1.0

This will force the user to pick one of the packages providing it, and Composer will suggest providers..

**Reply by sirmonti** [reply]
This works if your module works with existing objects, if your module wants to create objects you need to reference the concrete class. That is, the program must provide support for every real class.
Another way to do this is to provide a simple way to identify the real packages a virtual one has installed ;)
The only way I known to do that is to get the class list with the get_declared_classes() function, and scan which ones implements the desired interface.
One simple way is to add in autoload.php file a constant with an array that maps virtual packages to real ones.

**Reply by Seldaek** [reply]
Well, you can also rely on InstalledVersions https://getcomposer.org/doc/07-runtime.md#knowing-whether-package-x-or-virtual-package-is-present to figure out if something is there.

But yeah I see if you rely on concrete implementations and not just interfaces that's difficult. I guess for that relying on the php-http packages to abstract away the implementations might help.
